### PR TITLE
WIP: Maven Liquibase changelog generation for h2disk

### PR DESCRIPTION
**Pull request by Tcharl**
Upgrade liquibase version and fix liquibase diff for at least h2 db.
I will try to do the same for gradle soon :-)

- Please make sure the below checklist is followed for Pull Requests.

- [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [ ] Tests are added where necessary
- [ ] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by PierreBesson** [comment]
Wow if this really fixes liquibase:diff for H2, this is amazing !

**Comment by DanielFran** [comment]
@Tcharl I believe that the upgrade of dependencies should be done in jhipster/jhispter-depedencies project.

About liquibase update, there is some impacts in Spring that will be done only for next 2.1.0 M1 due to Liquibase changed logging to slf4j.

see my comments https://github.com/spring-projects/spring-boot/pull/13625#issuecomment-405653741

and the changes about logging: https://github.com/spring-projects/spring-boot/commit/d5e4a194461375b7bcd7432c29bdfea527a9d6a9

@jdubois @deepu105 Maybe it is better to start a new branch for migration to 2.1.0 like it was done for 2.0.x?

**Comment by Tcharl** [comment]
@DanielFran I'll try to see what can be put in jh-dependencies, but I think that at least liquibase-plugin version should be on the server generator side.
I don't think this is impacting the project slf4j version as it is within the liquibase plugin (not as a Maven dependencies).

@PierreBesson this is working, but liquibase + H2 as some issues:
* when generating diff, drop and create table are in uppercase instead of lower case (a fix could be to generate all table with uppercase as it is a well known convention).
* also, liquibase:status returns hash errors, even if there are not any.
* finally, generated diff adds a constraint for the jhi_user_authority which breaks oauth2, as the injected csvs are wrong: I'm gonna fix that  when I'll be sure that it is used nowhere (#8018).
But at least, it generates a functional changelog with real diff and not full database schema :-)

**Comment by DanielFran** [comment]
@Tcharl I was thinking that the PR was not finalized and you will also upgrade liquibase-core to last version ;)

If it is only the plugin, seems ok.

File: generators/server/templates/pom.xml.ejs [hunk 1]
```diff
@@ -958,6 +960,29 @@
...
                         <version>${validation-api.version}</version>
                     </dependency>
+                    <%_ if (devDatabaseType === 'h2Disk') { _%>
+                    <dependency>
+                        <groupId>com.h2database</groupId>
+                        <artifactId>h2</artifactId>
+                        <version>${h2.version}</version>
+                    </dependency>
+                    <%_ } _%>
+                    <dependency>
```
Comment (deepu105) [inline generators/server/templates/pom.xml.ejs]: why are these added?

**Comment by Tcharl** [comment]
It looks like the plug-in dependency is mandatory, at least for h2.

This or need more tests (I'll change it to wipe asap)

**Comment by deepu105** [comment]
@PierreBesson @pascalgrimaud can any of you check this?

**Comment by pascalgrimaud** [comment]
@deepu105 : ok, self assigned

File: generators/server/templates/pom.xml.ejs [hunk 2]
```diff
@@ -84,8 +84,10 @@
         <javassist.version>3.22.0-GA</javassist.version>
         <!-- The liquibase version should match the one managed by
         https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-dependencies/${spring-boot.version} -->
-        <liquibase.version>3.5.5</liquibase.version>
+        <liquibase.version>3.6.2</liquibase.version>
         <liquibase-hibernate5.version>3.6</liquibase-hibernate5.version>
+        <h2.version>1.4.197</h2.version>
```
Comment (pascalgrimaud) [inline generators/server/templates/pom.xml.ejs]: as only need by h2disk in dev, it could be with conditional: `<%_ if (devDatabaseType === 'h2Disk') { _%>`

**Comment by pascalgrimaud** [comment]
Probably not related to this PR, but `liquibase:diff` always generate.
Do someone understand why ?

```xml
    <changeSet author="pgrimaud (generated)" id="1533297526636-1">
        <dropUniqueConstraint constraintName="UC_JHI_USEREMAIL_COL" tableName="jhi_user"/>
        <addUniqueConstraint columnNames="email" constraintName="UC_JHI_USEREMAIL_COL" tableName="jhi_user"/>
    </changeSet>
    <changeSet author="pgrimaud (generated)" id="1533297526636-2">
        <dropUniqueConstraint constraintName="UC_JHI_USERLOGIN_COL" tableName="jhi_user"/>
        <addUniqueConstraint columnNames="login" constraintName="UC_JHI_USERLOGIN_COL" tableName="jhi_user"/>
    </changeSet>
```

**Comment by pascalgrimaud** [comment]
Excepting 1 minor comment, it looks like the same behaviour than MySQL, so better than what we currently have with H2Disk

**Comment by Tcharl** [comment]
Hi Pascal,

I've the same issue concerning the user's constraints, and the worse is that even with a right changelog, these constraints are generated everytime (and only these ones...).

Before merging, I would like to:

- upgrade the liquibase dependencies on the jhipster-dependencies pom

- check if for other databases their dependencies are not required (as the h2disk one here) (@deepu105 this is due to the upgrade, they restricted their dependencies, so we're constrained to set them up in the plugin configuration)

- Do the same on the gradle part

- Parametrize versions using pom properties

But I'm really under the wave for now, at least until september: we're finalizing a project at work, and then we don't know it's future, so it as to be pristine until then. And I'm really sorry for that.

**Comment by DanielFran** [comment]
@Tcharl if you will work on liquibase upgrade see my comments bellow...

**Comment by Tcharl** [comment]
Waiting for liquibase dependency upgrade by spring-boot-dependencies...

**Comment by DanielFran** [comment]
@Tcharl Can you review this PR?
Should be now possible to proceed with the fix....

**Comment by DanielFran** [comment]
In this PR we have 2 changes:
1 - migrate to use latest liquibase version that is already done in current master
2 - fix liquibase diff for H2 disk

For point 2 the changes I see in the code is:
before:
    if (devDatabaseType === 'h2Disk') { %>jdbc:h2:file:./**target**/h2db/db/<%= lowercaseBaseName %><% }

after:
if (devDatabaseType === 'h2Disk') { %>jdbc:h2:file:**${project.build.directory}**/h2db/db/<%= lowercaseBaseName %><% }

If I take a look in current pom in master we have:
<changeLogFile>**${project.basedir}/**<%= SERVER_MAIN_RES_DIR %>config/liquibase/master.xml</changeLogFile>
<diffChangeLogFile>**${project.basedir}**/<%= SERVER_MAIN_RES_DIR %>config/liquibase/changelog/${maven.build.timestamp}_changelog.xml</diffChangeLogFile>
...                        
<defaultSchemaName>... if (devDatabaseType === 'h2Disk') { %>jdbc:h2:file:**${project.build.directory}**/h2db/db/<%= lowercaseBaseName %><% } ...

So this seems that the fiz is already done!
@jhipster/developers can anybody confirm?

**Comment by DanielFran** [comment]
@jdubois @Tcharl I believe we can close this PR?

**Comment by jdubois** [comment]
@DanielFran I just tried it on master and I had a class not found exception, let me have a look

**Comment by jdubois** [comment]
I don't get it, now you need to add:

```
                        <dependency>
                            <groupId>org.springframework</groupId>
                            <artifactId>spring-beans</artifactId>
                            <version>5.1.4.RELEASE</version>
                        </dependency>
                        <dependency>
                            <groupId>org.springframework</groupId>
                            <artifactId>spring-core</artifactId>
                            <version>5.1.4.RELEASE</version>
                        </dependency>
                        <dependency>
                            <groupId>org.springframework</groupId>
                            <artifactId>spring-context</artifactId>
                            <version>5.1.4.RELEASE</version>
                        </dependency>
```

Otherwise the Maven Liquibase plugin has a classpath exception. I'm pretty sure it wasn't the case before.

As those plugins do not share the dependencies management from the parent pom (I wonder if/how that would be possible), an issue with that is that we would need to set up a specific variable for storing the Spring Framework version, like we already do for Spring Boot and Hibernate. So that's more maintenance.

**Comment by DanielFran** [comment]
@jdubois Do you want I or maybe @Tcharl create a new PR with that changes?

**Comment by jdubois** [comment]
Thanks @Tcharl yes if you have the time (as I know I won't).
As we have no parent pom (which is good, clients usually have their own), there's no other way to add a new variable for the Spring version - or do you have a better idea? This really sucks

**Comment by DanielFran** [comment]
@Tcharl Do you want to review the PR or close it and make a new one?

**Comment by Tcharl** [comment]
Hi,

Yes, I can update it with the new master branch.

@jdubois I think I know maven pretty well, but I have no magic stick. I'll see what I can do the less annoying way...

**Comment by jdubois** [comment]
Thanks @Tcharl - if that's the only way, then so be it!

**Comment by Tcharl** [comment]
Can close this one, I rebase in #9519
